CMakeLists.txt

```diff
@@ -1,227 +1,182 @@
+	set(cargo_toml_dir ${CMAKE_CURRENT_BINARY_DIR})
+	set(CARGO_PROJECT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/")
+	file(COPY 
+		${CARGO_PROJECT_DIR}/splitguide.yaml 
+		${CARGO_PROJECT_DIR}/cbindgen.toml
+		${CARGO_PROJECT_DIR}/rust-toolchain 
+		DESTINATION ${cargo_toml_dir})
+	set(cargo_generated_include_dir ${cargo_toml_dir}/include)
 endif()
+set(source_include_dir ${CMAKE_CURRENT_SOURCE_DIR}/include)
+set(cargo_target_dir ${cargo_toml_dir}/target)
+cmake_path(APPEND cargo_target_dir ${ZENOHC_CUSTOM_TARGET})
+set(cargo_binary_dir_debug ${cargo_target_dir}/debug)
+set(cargo_binary_dir_release ${cargo_target_dir}/release)
+set(cargo_binary_dir ${cargo_target_dir}/$<IF:$<CONFIG:Debug>,debug,release>)
 
-set(CARGO_FLAGS ${CARGO_FLAGS} "--manifest-path=${CMAKE_CURRENT_SOURCE_DIR}/Cargo.toml")
-set(RUSTFLAGS $ENV{RUSTFLAGS})
-add_custom_target(cargo ALL COMMAND cargo ${CARGO_CHANNEL} build ${CARGO_FLAGS} COMMENT "Running RUSTFLAGS=${RUSTFLAGS} cargo ${CARGO_CHANNEL} build ${CARGO_FLAGS}" WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}")
-
-#
-# Rust cross-build check for supported processor architectures
-# This check works on linux only
-# It requires that the following packages are installed for cross compilation:
 #
-# sudo apt install gcc-arm-linux-gnueabi
-# sudo apt install gcc-x86-64-linux-gnu
-# sudo apt install gcc-aarch64-linux-gnu
+# Configure Cargo.toml
 #
-# and the following targets in rustup
+set(CARGO_PROJECT_VERSION "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}")
+if(NOT PROJECT_VERSION_TWEAK)
+	set(CARGO_PROJECT_VERSION "${CARGO_PROJECT_VERSION}-dev")
+elseif(PROJECT_VERSION_TWEAK LESS 255)
+	set(CARGO_PROJECT_VERSION "${CARGO_PROJECT_VERSION}-rc.${PROJECT_VERSION_TWEAK}")
+endif()
+status_print(CARGO_PROJECT_VERSION)
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Cargo.toml.in" "${cargo_toml_dir}/Cargo.toml" @ONLY)
+
 #
-# rustup target add arm-unknown-linux-gnueabi
-# rustup target add aarch64-unknown-linux-gnu
-# rustup target add x86_64-unknown-linux-gnu
-# 
-# check which targets are already installed with
-# 
-# rustup target list --installed
+# Configure result library names
 #
-add_custom_target(crosscheck)
-
-set(targets
-aarch64-unknown-linux-gnu
-x86_64-unknown-linux-gnu
-arm-unknown-linux-gnueabi
-)
-
-foreach(target ${targets})
-	add_custom_target(cargo_check_${target} 
-		COMMAND cargo check --target ${target} ${CARGO_FLAGS} 
-		COMMENT "cargo check on ${target}" 
-		WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}")
-	add_dependencies(crosscheck cargo_check_${target})
-endforeach()
-
+macro(set_lib list var value)
+	set(${var} ${value}) 
+	list(APPEND ${list} ${value})
+endmacro()
+
+# dylib - dymamic library (.so, .dll, .dylib)
+# staticlib - static library (.a, .lib)
+# implib - import library for windows dynamic library (DLL) - .lib
+# dylibs - list of files required for use dynamic libraty
+# staticlibs - list of files required for use static libraty
 if(APPLE)
```

Comment:
All the platform-dependent switch below can be replaced by two lines:
set_lib(dylibs dylib "${CMAKE_SHARED_LIBRARY_PREFIX}zenohc${CMAKE_SHARED_LIBRARY_SUFFIX}")
set_lib(staticlibs staticlib "${CMAKE_STATIC_LIBRARY_PREFIX}zenohc${CMAKE_STATIC_LIBRARY_SUFFIX}")

Reply:
I won't say that if() elseif() elseif() is simpler. Why do we need *.dll.lib naming in Windows? Looks like smth redundant. Maybe it is an option to stay unified with *.lib instead and simplify everything?

Reply:
The zenohc.dll.lib is so-named 'import' library for corresponding DLL. Unlike linux or mac, where all the data for linker is stored in dynamic library (.so) itself, to link the windows .DLL the special static "import library" is required, which job is to load the DLL. CMake have support for this windows specific (IMPLIB parameters).

So the code will look like this:
```
set_lib(dylibs dylib "${CMAKE_SHARED_LIBRARY_PREFIX}zenohc${CMAKE_SHARED_LIBRARY_SUFFIX}")
set_lib(staticlibs staticlib "${CMAKE_STATIC_LIBRARY_PREFIX}zenohc${CMAKE_STATIC_LIBRARY_SUFFIX}")
if(WIN32)
  set_lib(dylibs dylib "${CMAKE_STATIC_LIBRARY_SUFFIX}zenohc${CMAKE_STATIC_LIBRARY_SUFFIX}${CMAKE_SHARED_LIBRARY_SUFFIX}")
endif()
```
which I think is quite ugly and strange